primitives/state-machine/src/ext.rs

```diff
@@ -109,6 +113,8 @@ where
 		changes_trie_state: Option<ChangesTrieState<'a, H, N>>,
 		extensions: Option<&'a mut Extensions>,
 	) -> Self {
+		overlay.enter_runtime()
```

Comment:
My reason for doing so is that `call_api_at` calls into state-machine which is still client code and starts additional transactions for some execution strategies.

That said, I think it would be OK to treat those transactions as runtime ones. Is there a specific reason you would attach that rather to `call_api_at` than to the ext?

Reply:
No it is mainly to avoid changing previous behavior.

It always felt like the management of prospective layer with the dual execution is a bit hacky so binding this to Ext lifetime can be good (yet it means that starting a transaction cannot really be bound to enter_runtime).
I do not have strong expectation of Ext usage (maybe there are), looks like a slim struct that intends to be instantiated in different contexts just to provide Externalities, I mean I would expect it to be ok to build in any circumstance, but it can be fine to attach this info to its lifetime too.

Reply:
As a matter of fact some benchmark fails because they double exiting the the runtime. I guess because they are doing weird stuff with the `Ext`. Maybe we should attach it to `call_api_at` after all.

Reply:
Would need to bring back the 'hacky' two execution strategy code in some way (probably calling runtime_enter and forcing back the previous state on exit).
But iirc there is a test that is triggered when things are not working properly there (not a state-machine one).
Not sure there is a perfect solution there.

Reply:
So I moved enter/exit to `call_api_at` like this:

```
self.changes.borrow_mut().enter_runtime().unwrap();
let res = call_api_at(
	&self.call,
	self,
	&self.changes,
	&self.offchain_changes,
	&self.storage_transaction_cache,
	&self.initialized_block,
	&self.recorder,
);
self.changes.borrow_mut().exit_runtime().unwrap();
```


```
 0: sp_panic_handler::set::{{closure}}
   1: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
   2: rust_begin_unwind
             at src/libstd/panicking.rs:378
   3: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   4: core::option::expect_none_failed
             at src/libcore/option.rs:1211
   5: <node_runtime::RuntimeApiImpl<__SR_API_BLOCK__,RuntimeApiImplCall> as sp_api::Core<__SR_API_BLOCK__>>::Core_initialize_block_runtime_api_impl
   6: sp_api::Core::initialize_block
   7: <sc_service::client::call_executor::LocalCallExecutor<B,E> as sc_client_api::call_executor::CallExecutor<Block>>::contextual_call
   8: <sc_service::client::client::Client<B,E,Block,RA> as sp_api::CallApiAt<Block>>::call_api_at
   9: sp_consensus_babe::runtime_decl_for_BabeApi::configuration_call_api_at
  10: <node_runtime::RuntimeApiImpl<__SR_API_BLOCK__,RuntimeApiImplCall> as sp_consensus_babe::BabeApi<__SR_API_BLOCK__>>::BabeApi_configuration_runtime_api_impl
  11: sp_consensus_babe::BabeApi::configuration
  12: sc_consensus_slots::SlotDuration<T>::get_or_compute
  13: sc_consensus_babe::Config::get_or_compute
  14: sc_service::builder::ServiceBuilder<TBl,TRtApi,TCl,TFchr,TSc,TImpQu,TFprb,TFpp,TExPool,TRpc,Backend>::with_import_queue
  15: node_cli::service::new_full
  16: sc_cli::runner::Runner<C>::run_node
  17: node_cli::command::run
  18: substrate::main
  19: std::rt::lang_start::{{closure}}
  20: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
      std::panicking::try::do_call
             at src/libstd/panicking.rs:303
  21: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:86
  22: std::panicking::try
             at src/libstd/panicking.rs:281
      std::panic::catch_unwind
             at src/libstd/panic.rs:394
      std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  23: main
  24: __libc_start_main
  25: _start


Thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: AlreadyInRuntime', /home/athei/remote-builds/17179518214072101788/bin/node/runtime/src/lib.rs:890

test check_block_works ... FAILED
```

Reply:
Oh, I was more thinking of putting it in 'impl_runtime_api' (where tx are started and in commit_on_ok function). Sticking to the old commit/discard prospective calls.

Reply:
> Oh, I was more thinking of putting it in 'impl_runtime_api' (where tx are started and in commit_on_ok function)

This is exactly where this code snippet is from. 

> Edit: with the existing associated boolean logic to avoid recursively calling it

This is what I am trying to understand. Why is that even called recursively? We assert on purpose that it is NOT called recursively. If it is, that is the wring place to put it.

Reply:
I moved it to the state-machine now. Which seems to be the right place as it is not called recursively in my first test. Let's wait what CI says.

Reply:
Yes that is clearly some part of the code that is a bit opaque, and one of the reason I tend to prefer the conservative approach.
My understanding is that the previous commited/prospective is mainly needed to include or not a transaction only while building block.
So the real position could be there, but I wouldn't bet to much on it.

Reply:
If you are strictly targetting runtime call boundary it may be the right location (call executor spawn a state machine on every call), but I would fear in this case we are a bit over calling it. I mean we change the way thing works, could be ok though.

Reply:
Just read your code, I was wrongly thinking you did wrap the the enter/exit runtime over the lifetime of the state machine struct, actually it is wrapping 'execute_aux' which looks good to me.